### PR TITLE
Stage securityagent integration for release

### PR DESCRIPTION
**Pull request by nr-swilloughby**
This merges the nrsecurityagent into the develop branch to complete testing in preparation for the next release of the go agent.

### CODE REVIEW

File: v3/integrations/nrmicro/go.mod [hunk 1]
```diff
@@ -2,10 +2,12 @@ module github.com/newrelic/go-agent/v3/integrations/nrmicro
...
 
 require (
-	github.com/golang/protobuf v1.3.2
+	github.com/golang/protobuf v1.5.3
 	github.com/micro/go-micro v1.8.0
-	github.com/newrelic/go-agent/v3 v3.4.0
+	github.com/newrelic/go-agent/v3 v3.22.0
 )
+
+replace github.com/newrelic/go-agent/v3 v3.22.0 => ../..
```
Comment (iamemilio) [inline v3/integrations/nrmicro/go.mod:13]:
```suggestion

```

This requires the go agent to be in the local path at ../..

File: v3/integrations/nrmongo/go.mod [hunk 2]
```diff
@@ -2,10 +2,12 @@ module github.com/newrelic/go-agent/v3/integrations/nrmongo
...
+go 1.18
 
 require (
-	github.com/newrelic/go-agent/v3 v3.18.2
+	github.com/newrelic/go-agent/v3 v3.22.0
 	// mongo-driver does not support modules as of Nov 2019.
 	go.mongodb.org/mongo-driver v1.10.2
 )
+
+replace github.com/newrelic/go-agent/v3 v3.22.0 => ../..
```
Comment (iamemilio) [inline v3/integrations/nrmongo/go.mod:13]:
```suggestion

```

File: v3/integrations/nrsecurityagent/README.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
+# v3/integrations/nrsecurityagent [![GoDoc](https://godoc.org/github.com/newrelic/go-agent/v3/integrations/nrsecurityagent?status.svg)](https://godoc.org/github.com/newrelic/go-agent/v3/integrations/nrsecurityagent)
+
+This integration allows you to have the New Relic security agent analyze your application for potentially exploitable vulnerabilities.
```
Comment (iamemilio) [inline v3/integrations/nrsecurityagent/README.md:3]:
```suggestion
The New Relic security agent analyzes your application for potentially exploitable vulnerabilities.
```

File: v3/integrations/nrsecurityagent/go.mod [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+	github.com/newrelic/csec-go-agent v0.2.1
+	github.com/newrelic/go-agent/v3 v3.22.0
+	github.com/newrelic/go-agent/v3/integrations/nrsqlite3 v1.1.1
+	gopkg.in/yaml.v2 v2.4.0
+)
+
+replace (
+	github.com/newrelic/go-agent/v3 v3.22.0 => ../..
+	github.com/newrelic/go-agent/v3/integrations/nrsqlite3 => ../nrsqlite3
+)
```
Comment (iamemilio) [inline v3/integrations/nrsecurityagent/go.mod:15]:
```suggestion

```

**Comment by iamemilio** [comment]
I am a little concerned about the logrus test failing too

**Comment by nr-swilloughby** [comment]
Thanks for catching those vestiges of the prerelease code still in there. They'd have been cleaned up during release prep but it's better to get them now.

File: v3/integrations/nrgrpc/nrgrpc_server.go [hunk 5]
```diff
@@ -306,8 +307,9 @@ func UnaryServerInterceptor(app *newrelic.Application, options ...HandlerOption)
 		option(localHandlerMap)
 	}
 
-	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+	return func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp any, err error) {
 		txn := startTransaction(ctx, app, info.FullMethod)
+		newrelic.GetSecurityAgentInterface().SendEvent("GRPC", req)
```
Comment (mirackara) [inline v3/integrations/nrgrpc/nrgrpc_server.go:312]: Are we still following the naming convention for gRPC? 
"GRPC" -> "gRPC"

Comment (mirackara) [inline v3/integrations/nrgrpc/nrgrpc_server.go:312]:
```suggestion
		newrelic.GetSecurityAgentInterface().SendEvent("gRPC", req)
```

File: v3/integrations/nrgrpc/nrgrpc_server.go [hunk 6]
```diff
@@ -327,6 +329,11 @@ func (s wrappedServerStream) Context() context.Context {
 	return newrelic.NewContext(ctx, s.txn)
 }
 
+func (s wrappedServerStream) RecvMsg(msg any) error {
+	newrelic.GetSecurityAgentInterface().SendEvent("GRPC", msg)
```
Comment (mirackara) [inline v3/integrations/nrgrpc/nrgrpc_server.go:333]:
```suggestion
	newrelic.GetSecurityAgentInterface().SendEvent("gRPC", msg)
```

File: v3/integrations/nrgrpc/nrgrpc_server.go (hunk 5 above)
Comment (nr-swilloughby) [reply v3/integrations/nrgrpc/nrgrpc_server.go:312]: That's what the security agent expects (this is an internal protocol between the agents. We can't just change it on our end.

File: v3/integrations/nrb3/go.mod [hunk 7]
```diff
@@ -1 +1,5 @@
 module github.com/newrelic/go-agent/v3/integrations/nrb3
+
+go 1.20
```
Comment (iamemilio) [inline v3/integrations/nrb3/go.mod:3]:
```suggestion
go 1.20
```

does this need to be Golang version 1.20?

Comment (nr-swilloughby) [reply v3/integrations/nrb3/go.mod:3]: I don't know. nrb3 isn't part of the work done for this. nothing changed except the go.mod file, probably by a script while cleaning up go.mods. I don't have a problem moving it back to the lowest supported release.

File: v3/integrations/nrmongo/go.mod [hunk 8]
```diff
@@ -2,10 +2,11 @@ module github.com/newrelic/go-agent/v3/integrations/nrmongo
 
 // As of Dec 2019, 1.10 is the mongo-driver requirement:
 // https://github.com/mongodb/mongo-go-driver#requirements
-go 1.17
+go 1.18
 
 require (
-	github.com/newrelic/go-agent/v3 v3.18.2
+	github.com/newrelic/go-agent/v3 v3.22.0
```
Comment (iamemilio) [inline v3/integrations/nrmongo/go.mod:8]:
```suggestion
	github.com/newrelic/go-agent/v3 v3.22.1
```

We probably should not require the version of the agent we recently dropped from the go eco system.

File: v3/integrations/nrgin/go.mod [hunk 9]
```diff
@@ -2,9 +2,9 @@ module github.com/newrelic/go-agent/v3/integrations/nrgin
 
 // As of Dec 2019, the gin go.mod file uses 1.12:
 // https://github.com/gin-gonic/gin/blob/master/go.mod
-go 1.12
+go 1.18
 
 require (
 	github.com/gin-gonic/gin v1.9.0
-	github.com/newrelic/go-agent/v3 v3.18.2
+	github.com/newrelic/go-agent/v3 v3.22.0
```
Comment (iamemilio) [inline v3/integrations/nrgin/go.mod:9]:
```suggestion
	github.com/newrelic/go-agent/v3 v3.22.1
```

File: v3/integrations/nrgrpc/go.mod [hunk 10]
```diff
@@ -1,22 +1,12 @@
...
-go 1.17
+go 1.18
 
 require (
 	// protobuf v1.3.0 is the earliest version using modules, we use v1.3.1
 	// because all dependencies were removed in this version.
-	github.com/golang/protobuf v1.5.2
-	github.com/newrelic/go-agent/v3 v3.18.2
+	github.com/golang/protobuf v1.5.3
+	github.com/newrelic/go-agent/v3 v3.22.0
```
Comment (iamemilio) [inline v3/integrations/nrgrpc/go.mod:9]:
```suggestion
	github.com/newrelic/go-agent/v3 v3.22.1
```

**Review by mirackara — APPROVED** [review]

**Pull request merged**